src/main/java/com/minecolonies/coremod/entity/ai/citizen/herders/AbstractEntityAIHerder.java

```diff
@@ -0,0 +1,534 @@
+    {
+        toolsNeeded.add(ToolType.AXE);
+        itemsNeeded.add(new ItemStack(getBreedingItem(), 2));
+
+        for (final ToolType tool : toolsNeeded)
+        {
+            if (checkForToolOrWeapon(tool))
+            {
+                return getState();
+            }
+        }
+
+        for (final ItemStack item : itemsNeeded)
+        {
+            checkOrRequestItemsAsynch(false, item);
+        }
+
+        return HERDER_DECIDE;
+    }
+
+    /**
+     * Butcher some animals (Preferably Adults) that the herder looks after.
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState butcherAnimals()
+    {
+        setDelay(DELAY_TWENTY);
+
+        if (!maxAnimals())
+        {
+            return HERDER_DECIDE;
+        }
+
+        if (!equipTool(ToolType.AXE))
+        {
+            return START_WORKING;
+        }
+
+        EntityAnimal animal = searchForAnimals().stream().filter(animalToButcher -> !animalToButcher.isChild()).findFirst().orElse(null);
+
+        if (animal == null)
+        {
+            animal = searchForAnimals().stream().findFirst().orElse(null);
+        }
+
+        butcherAnimal(animal);
+
+        if (animal != null && !animal.isEntityAlive())
+        {
+            incrementActionsDone();
+        }
+
+        return HERDER_BUTCHER;
+    }
+
+    /**
+     * Breed some animals together.
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState breedAnimals()
+    {
+        setDelay(DELAY_FOURTY);
+
+        final List<T> animals = new ArrayList<>(searchForAnimals());
+
+        final EntityAnimal animalOne = animals.stream().filter(animal -> animal.getGrowingAge() == 0).findAny().orElse(null);
+
+        if (animalOne == null)
+        {
+            return HERDER_DECIDE;
+        }
+
+        final EntityAnimal animalTwo = animals.stream().filter(animal ->
+          {
+              final float range = animal.getDistanceToEntity(animalOne);
+              final boolean isAnimalOne = animalOne.equals(animal);
+              return animal.getGrowingAge() == 0 && range <= DISTANCE_TO_BREED && !isAnimalOne;
+          }
+        ).findAny().orElse(null);
+
+        if (animalTwo == null)
+        {
+            return HERDER_DECIDE;
+        }
+
+        if (!equipItem(new ItemStack(getBreedingItem(), 2)))
+        {
+            return START_WORKING;
+        }
+
+        worker.setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.herder.breeding"));
+
+        breedTwoAnimals(animalOne, animalTwo);
+
+        return HERDER_DECIDE;
+    }
+
+    /**
+     * Allows the worker to pickup any stray items around Hut.
+     * Specifically useful when he possibly leaves Butchered
+     * drops OR with chickens (that drop feathers and etc)!
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState pickupItems()
+    {
+        final List<EntityItem> items = new ArrayList<>(searchForItemsInArea());
```

Comment:
no need to double copy :)

Reply:
for some reason Sonar/IntelliJ inspections didn't like it. but i can just shut them up :smile: